### PR TITLE
Refactor Upstream API for setting TCP tunneling mode

### PR DESCRIPTION
**Pull request by yanavlasov**
Commit message:
This API was introduced by #32991 and it conflates support for half-close with TCP proxying mode in H/1 codec. This interferes with #34461 which enables support for server half-close in H/2 proxying.

This PR refactors API to separate enabling of the half-close behavior from TCP proxy behavior. Prior to #34461 half-close semativs were only enabled for TCP proxy upstreams, but #34461 introduces other cases as well. This necessitates setting half-close and TCP proxy modes separately. TCP proxy mode is now enabled in the class dedicated to the TCP proxy upstreams.

Risk Level: Low
Testing: Unit Tests (with runtime flag enabled)
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Review by alyssawilk — COMMENTED** [review]
/wait-any

File: source/common/router/upstream_request.cc
Comment (yanavlasov) [reply source/common/router/upstream_request.cc:601]: The `enable_half_close_` value is plumbed through the `FilterManagerCallbacks::isHalfCloseEnabled()` to the filter manager, which determines if `local_complete_` is set on encoder end_stream for the upstream filter manager.

The `GenericUpstream::enableHalfClose()` is setting the TCP proxying mode in H/1 codec which controls how codec behaves when it sees decoder end stream. This code used the half-close flag here to also set the TCP proxying mode, which worked before since this was the half-close mode was only used in TCP proxy mode.

In #34461 I'm going to enable half-close mode in other cases, so the TCP proxy settings needs to be separate from half-close. So I have moved setting the TCP proxy mode to the dedicated `RouterUpstreamRequest` class for TCP proxy upstreams.

I have updated PR description accordingly.

File: source/common/tcp_proxy/upstream.h
Comment (alyssawilk) [inline source/common/tcp_proxy/upstream.h:62]: class comment please

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo a class comment explaining why the subclass and when it was used.  I'm fine with any maintainer sign-off on the comment so feel free to have someone from EPT review when I'm out tomorrow to improve velocity.

File: source/common/tcp_proxy/upstream.h
Comment (yanavlasov) [reply source/common/tcp_proxy/upstream.h:65]: Done

**Review by RyanTheOptimist — APPROVED** [review]
Thanks for cleaning this up!

**Pull request merged**
